upload/upload.go

```diff
@@ -0,0 +1,201 @@
+package upload
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pinpt/go-common/api"
+	"github.com/pinpt/httpclient"
+)
+
+// NOTE: these defaults are borrowed from https://github.com/aws/aws-sdk-go/blob/master/service/s3/s3manager/upload.go
+
+// MaxUploadParts is the maximum allowed number of parts in a multi-part upload
+// on Amazon S3.
+const MaxUploadParts = 10000
+
+// MinUploadPartSize is the minimum allowed part size when uploading a part to
+// Amazon S3.
+const MinUploadPartSize int64 = 1024 * 1024 * 5
+
+// DefaultUploadPartSize is the default part size to buffer chunks of a
+// payload into.
+const DefaultUploadPartSize = MinUploadPartSize
+
+// DefaultUploadConcurrency is the default number of goroutines to spin up when
+// using Upload().
+const DefaultUploadConcurrency = 5
+
+// MaxAttemptsForPart is the max number of attempts for a part before failing
+const MaxAttemptsForPart = 10
+
+// Options for configuring the upload
+type Options struct {
+	// The buffer size (in bytes) to use when buffering data into chunks and
+	// sending them as parts to S3. The minimum allowed part size is 5MB, and
+	// if this value is set to zero, the DefaultUploadPartSize value will be used.
+	PartSize int64
+
+	// The number of goroutines to spin up in parallel per call to Upload when
+	// sending parts. If this is set to zero, the DefaultUploadConcurrency value
+	// will be used.
+	//
+	// The concurrency pool is not shared between calls to Upload.
+	Concurrency int
+
+	// AttemptsForPart is the max number of attempts to retry (if failed) before
+	// given up. Defaults to MaxAttemptsForPart
+	AttemptsForPart int
+
+	// Cookies is the cookie headers that must be set to provide permission
+	// for uploading
+	Cookies []string
+
+	// Body is the content to upload
+	Body io.ReadCloser
+
+	// ContentType is the body content type
+	ContentType string
+
+	// URL to upload the parts to
+	URL string
+
+	// Job information about the upload which will be saved into a file named job.json in the same folder
+	Job map[string]interface{}
+}
+
+type part struct {
+	index  int
+	reader io.Reader
+}
+
+func newClient() (httpclient.Client, error) {
+	hcConfig := &httpclient.Config{
+		Paginator: httpclient.NoPaginator(),
+		Retryable: httpclient.NewBackoffRetry(10*time.Millisecond, 100*time.Millisecond, 30*time.Second, 2.0),
+	}
+	return api.NewHTTPAPIClient(hcConfig)
```

Comment:
Should probably make it possible to pass Client into Options. For example when just testing upload, we may want to lower retries so it takes faster to validate.

Reply:
good idea. will add